app/models/manageiq/providers/cisco_intersight/inventory/collector/physical_infra_manager.rb

```diff
@@ -3,133 +3,147 @@ class Inventory::Collector::PhysicalInfraManager < ManageIQ::Providers::Inventor
 
     def collect
 
-      set_configuration
-
+      # Establish connection. Connection is inside ManagerMixin which sets API key and keyid-d
+      connection
+
+      # Initialize API endpoints
+      firmware_api
+      compute_api
+      equipment_api
+      asset_api
+      adapter_api
+      management_api
+      storage_api
+      network_api
+      port_api
+      ether_api
+
+      # Initialize the variables that may be memoized for the duration of the refresh run
       physical_servers
       physical_racks
-      physical_server_network_devices
-      firmware_inventory
-
+      firmware_firmware_summaries
+      network_elements
+      physical_chassis
     end
 
-    def set_configuration
-      @connection ||= manager.connect
-    end
+    # Methods that are directly used by parser
 
-    def get_firmware_api
-      IntersightClient::FirmwareApi.new
+    def device_contract_informations
+      # Returns an array with objects of type DeviceContractInformation
+      @device_contract_informations ||= asset_api.get_asset_device_contract_information_list.results
     end
 
-    def get_compute_api
-      IntersightClient::ComputeApi.new
+    def physical_racks
+      @physical_racks ||= compute_api.get_compute_rack_unit_list.results
     end
 
-    def get_equipment_api
-      IntersightClient::EquipmentApi.new
+    def firmware_firmware_summaries
+      @firmware_firmware_summaries ||= firmware_api.get_firmware_firmware_summary_list.results
     end
 
-    def get_asset_api
-      IntersightClient::AssetApi.new
+    def physical_servers
+      @physical_servers ||= compute_api.get_compute_physical_summary_list.results
     end
 
-    def get_memory_api
-      IntersightClient::MemoryApi.new
+    def physical_chassis
+      @physical_chassis ||= equipment_api.get_equipment_chassis_list.results
     end
 
-    def get_adapter_api
-      IntersightClient::AdapterApi.new
+    def network_elements
+      @network_elements ||= network_api.get_network_element_list.results
     end
 
-    def get_management_api
-      IntersightClient::ManagementApi.new
+    def get_device_contract_information_from_device_moid(registered_device_moid)
+      device_contract_informations.find { |c| c.registered_device.moid == registered_device_moid }
```

Comment:
Minor but this can be a bit more efficient if you build a hash once and cache it instead of looping through the list every time, typically we'll do something like:
```ruby
def device_contract_informations_by_moid
  @device_contract_informations_by_moid ||= device_contract_informations.index_by(&:moid)
end
```

Then in your parser you would call `device_contract_informations_by_moid[device.moid]`

Good for a follow-up though not going to hold this up

Reply:
Ok, thanks for the response. Going to resolve that in the follow-up PR.